### PR TITLE
Feature JSON Pointers

### PR DESCRIPTION
**Pull request by jcdickinson**
Adds JSON pointers. These conform to [RFC 6901.](https://tools.ietf.org/html/rfc6901)

### CODE REVIEW

File: src/SourceCode.Clay.Json/Pointers/JsonPointer.cs [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        /// <summary>Indicates whether the current object is equal to another object of the same type.</summary>
+        /// <param name="other">An object to compare with this object.</param>
+        /// <returns>true if the current object is equal to the <paramref name="other">other</paramref> parameter; otherwise, false.</returns>
+        public bool Equals(JsonPointer other)
+        {
+            if (ReferenceEquals(_tokens, other._tokens)) return true;
+            if (ReferenceEquals(_tokens, null) || ReferenceEquals(other._tokens, null)) return false;
+            if (_tokens.Length != other._tokens.Length) return false;
+
+            for (var i = 0; i < _tokens.Length; i++)
```
Comment (grant-d) [inline src/SourceCode.Clay.Json/Pointers/JsonPointer.cs]: Consider Clay.Collections.Generic.ListEquals

File: src/SourceCode.Clay.Json/Pointers/JsonPointer.cs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+            }
+
+            return true;
+        }
+
+        /// <summary>Returns the hash code for this instance.</summary>
+        /// <returns>A 32-bit signed integer that is the hash code for this instance.</returns>
+        public override int GetHashCode()
+        {
+            var hashCode = -1248728428;
```
Comment (grant-d) [inline src/SourceCode.Clay.Json/Pointers/JsonPointer.cs]: add `unchecked` scope. Even if you don't need it now, someone maintaining the code may do different math

File: src/SourceCode.Clay.Json/Pointers/JsonPointer.cs (hunk 1 above)
Comment (jcdickinson) [reply src/SourceCode.Clay.Json/Pointers/JsonPointer.cs]: Don't want to add a reference for a single extension method.

File: src/SourceCode.Clay.Json/Pointers/JsonPointer.cs [hunk 3]
```diff
@@ -280,9 +282,12 @@ public bool Equals(JsonPointer other)
         /// <returns>A 32-bit signed integer that is the hash code for this instance.</returns>
         public override int GetHashCode()
```
Comment (grant-d) [inline src/SourceCode.Clay.Json/Pointers/JsonPointer.cs]: Could you use this pattern from CoreFx:
```
var hc = 17L;

            unchecked
            {
                hc = hc * 21 + _typeCode;

                
            }

            return ((int)(hc >> 32)) ^ (int)hc;
```

File: src/SourceCode.Clay.Json/Pointers/JsonPointerEvaluationOptions.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
+﻿using System;
+
+namespace SourceCode.Clay.Json.Pointers
+{
+    /// <summary>
+    /// Determines options for evaluating a <see cref="JsonPointer"/>.
+    /// </summary>
+    [Flags]
+    public enum JsonPointerEvaluationOptions
```
Comment (grant-d) [inline src/SourceCode.Clay.Json/Pointers/JsonPointerEvaluationOptions.cs]: Optional: `enum ... : byte`

**Pull request merged**